File: rust/day_4/part_1.rs
use std::fs;

fn solve(input: &str) -> i32 {
    let grid: Vec<&str> = input.trim().lines().collect();
    let rows = grid.len();
    let cols = if rows > 0 { grid[0].len() } else { 0 };

    let directions = [
        (-1, -1),
        (-1, 0),
        (-1, 1),
        (0, -1),
        (0, 1),
        (1, -1),
        (1, 0),
        (1, 1),
    ];

    let mut accessible_count = 0;

    for row in 0..rows {
        let row_chars: Vec<char> = grid[row].chars().collect();
        for col in 0..cols {
            if row_chars[col] == '@' {
                let mut adjacent_rolls = 0;

                for (dr, dc) in directions.iter() {
                    let new_row = row as i32 + dr;
                    let new_col = col as i32 + dc;

                    if new_row >= 0
                        && new_row < rows as i32
                        && new_col >= 0
                        && new_col < cols as i32
                    {
                        let new_row_chars: Vec<char> = grid[new_row as usize].chars().collect();
                        if new_row_chars[new_col as usize] == '@' {
                            adjacent_rolls += 1;
                        }
                    }
                }

                if adjacent_rolls < 4 {
                    accessible_count += 1;
                }
            }
        }
    }

    accessible_count
}

fn main() {
    let input = fs::read_to_string("../inputs/day_4.txt").expect("Failed to read input file");
    let result = solve(&input);
    println!("{}", result);
}

File: java/day_7/Part1.java
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

public class Part1 {
    public static int solve(String input) {
        String[] lines = input.trim().split("\n");
        List<char[]> grid = new ArrayList<>();
        for (String line : lines) {
            grid.add(line.toCharArray());
        }
        int rows = grid.size();
        int cols = rows > 0 ? grid.get(0).length : 0;

        int start_col = -1;
        for (int col = 0; col < cols; col++) {
            if (grid.get(0)[col] == 'S') {
                start_col = col;
                break;
            }
        }

        Set<Integer> beams = new HashSet<>();
        beams.add(start_col);
        int split_count = 0;

        for (int row = 1; row < rows; row++) {
            Set<Integer> new_beams = new HashSet<>();

            for (int col : beams) {
                if (col >= 0 && col < cols) {
                    if (grid.get(row)[col] == '^') {
                        split_count++;
                        if (col - 1 >= 0) {
                            new_beams.add(col - 1);
                        }
                        if (col + 1 < cols) {
                            new_beams.add(col + 1);
                        }
                    } else {
                        new_beams.add(col);
                    }
                }
            }

            beams = new_beams;
        }

        return split_count;
    }

    public static void main(String[] args) throws IOException {
        String input = new String(Files.readAllBytes(Paths.get("../inputs/day_7.txt")));
        int result = solve(input);
        System.out.println(result);
    }
}
